torch/csrc/jit/script/sugared_value.cpp

```diff
@@ -259,6 +259,135 @@ std::shared_ptr<SugaredValue> SimpleValue::call(
   return SugaredValue::call(loc, m, inputs, attributes, n_binders);
 }
 
+Value* SimpleValue::len(const SourceRange& loc, Function& m) {
+  // List, Tuple, Tensor, fill in missing information desugaring
+  Value* val = getValue();
+  TypePtr val_type = val->type();
+  Graph& g = *m.graph();
+  if (val_type->cast<ListType>() ||
+      val_type->isSubtypeOf(TensorType::get())) {
+    return g.insert(aten::len, {val}, {}, loc);
+  } else {
+    throw ErrorReport(loc)
+      << "cannot get length of the value type " << val_type->python_str();
+  }
+}
+
+Value* SimpleValue::getelem(const SourceRange&loc, Function& m, Value* i) {
+  Value* val = getValue();
+  TypePtr val_type = val->type();
+  Graph& g = *m.graph();
+  Value* cur_elem = nullptr;
+  if (val_type->cast<ListType>()) {
+    cur_elem = g.insert(aten::select, {val, i}, {}, loc);
+  } else if (val_type->isSubtypeOf(TensorType::get())) {
+    cur_elem = g.insert(aten::select, {val, 0, i}, {}, loc);
+  } else {
+    throw ErrorReport(loc)
+      << "cannot get element of the value type " << val_type->python_str();
+  }
+  return cur_elem;
+}
+
+RangeValue::RangeValue(const SourceRange& loc, Function& m, std::vector<Value*> inputs) {
+  Graph& g = *m.graph();
+  if(inputs.size() == 0) {
+    throw ErrorReport(loc) << "range expected at least 1 arguments, got 0";
+  } else if (inputs.size() == 1) {
+      end_ = inputs[0];
+      start_ = g.insertConstant(0, nullptr, loc);
+      step_ = g.insertConstant(1, nullptr, loc);
+      // range() call only contains end, easier to calculate len() and getitem()
+      is_simple_ = true;
+    } else if (inputs.size() <= 3) {
+      start_ = inputs[0];
+      end_ = inputs[1];
+      if (inputs.size() == 3) {
+        step_ = inputs[2];
+        // error handling when step_val = 0 during runtime
```

Comment:
Can we make __range_length just do this error handling? It is much more inline with how we already do error handling everywhere else.